t/mojo/template.t

```diff
@@ -20,170 +20,202 @@ sub exception { die 'ohoh' }
+  my $output = $mt->render(" bar\n foo\n    <%= 'test' =%>\n foo\n bar\n");
+  is $output, " bar\n foo\ntest foo\n bar\n", 'expression trimmed';
+};
+
+subtest 'Trim expression (multiple empty lines)' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render("    \n<%= 'test' =%>\n    ");
+  is $output, "    \ntest    \n", 'expression trimmed';
+};
+
+subtest 'Trim expression tags' => sub {
+  my $capture = 'no warnings "redefine"; sub capture { shift->(@_) }';
+  my $mt      = Mojo::Template->new(prepend => $capture);
+  my $output  = $mt->render('    <%= capture begin =%><html><% end =%>    ');
+  is $output, '<html>', 'expression tags trimmed';
+};
+
+subtest 'Trim expression tags (relaxed expression end)' => sub {
+  my $capture = 'no warnings "redefine"; sub capture { shift->(@_) }';
+  my $mt      = Mojo::Template->new(prepend => $capture);
+  my $output  = $mt->render('    <%= capture begin =%><html><%= end =%>    ');
+  is $output, '<html>', 'expression tags trimmed';
+};
+
+subtest 'Trim expression tags (relaxed escaped expression end)' => sub {
+  my $capture = 'no warnings "redefine"; sub capture { shift->(@_) }';
+  my $mt      = Mojo::Template->new(prepend => $capture);
+  my $output  = $mt->render('    <%= capture begin =%><html><%== end =%>    ');
+  is $output, '<html>', 'expression tags trimmed';
+};
+
+subtest 'Trim expression tags (trim reset)' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render('    <%= "one" =%><%= "two" %>  three');
+  is $output, "onetwo  three\n", 'expression tags trimmed';
+};
+
+subtest 'Nothing to trim' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render('<% =%>');
+  is $output, '', 'nothing trimmed';
+};
+
+subtest 'Replace tag' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render('<%% 1 + 1 %>');
+  is $output, "<% 1 + 1 %>\n", 'tag has been replaced';
+};
+
+subtest 'Replace expression tag' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render('<%%= 1 + 1 %>');
+  is $output, "<%= 1 + 1 %>\n", 'expression tag has been replaced';
+};
+
+subtest 'Replace expression tag (alternative)' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render(' lalala <%%= 1 + 1 %> 1234 ');
+  is $output, " lalala <%= 1 + 1 %> 1234 \n", 'expression tag has been replaced';
+};
+
+subtest 'Replace expression tag (another alternative)' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render(<<EOF);
 lalala <%%= 1 +
  1 %> 12
 34
 EOF
-is $output, "lalala <%= 1 +\n 1 %> 12\n34\n", 'expression tag has been replaced';
+  is $output, "lalala <%= 1 +\n 1 %> 12\n34\n", 'expression tag has been replaced';
+};
 
-# Replace comment tag
-$mt     = Mojo::Template->new;
-$output = $mt->render('<%%# 1 + 1 %>');
-is $output, "<%# 1 + 1 %>\n", 'comment tag has been replaced';
+subtest 'Replace comment tag' => sub {
+  my $mt     = Mojo::Template->new;
+  my $output = $mt->render('<%%# \'1 + 1 %>');
```

Comment:
Addition of `'` to a test input and expected response.  Test still passes, but unsure of the reason.

Reply:
I can undo this unintentional *adds*. To be honest, to do this restructure of the tests is a kind of monkey job. So I tried to do it with some *clever* regex substitutions, seems that went not that well ...